PyPI_analysis/detect_imports.py

```diff
@@ -214,75 +200,101 @@ def dynamic_imports(parsed_code: ast.Module):
+                                            ),
+                                        )
+                                    }
+                        elif isinstance(node_import, ast.ImportFrom):
+                            # Relative imports are always relative to the current package, and
+                            # will therefore not resolve to a third-party package.
+                            # They are therefore uninteresting to us.
+                            if (
+                                node_import.level == 0
+                                and node_import.module is not None
+                            ):
+                                name = node_import.module.split(".", 1)[0]
+                                if is_external_import(name):
+                                    yield {
+                                        "Docstring": ParsedImport(
+                                            name=name,
+                                            source=source.supply(
+                                                lineno=node.body[0].lineno
+                                            ),
+                                        )
+                                    }
+
+    def regular_imports(node: ast.AST):
+        if isinstance(node, ast.Import):
+            logger.debug(ast.dump(node))
+            for alias in node.names:
+                name = alias.name.split(".", 1)[0]
+                if is_external_import(name):
+                    yield {
+                        "Regular": ParsedImport(
+                            name=name, source=source.supply(lineno=node.lineno)
+                        )
+                    }
+        elif isinstance(node, ast.ImportFrom):
+            logger.debug(ast.dump(node))
+            # Relative imports are always relative to the current package, and
+            # will therefore not resolve to a third-party package.
+            # They are therefore uninteresting to us.
+            if node.level == 0 and node.module is not None:
+                name = node.module.split(".", 1)[0]
+                if is_external_import(name):
+                    yield {
+                        "Regular": ParsedImport(
+                            name=name, source=source.supply(lineno=node.lineno)
+                        )
+                    }
 
     try:
         parsed_code = ast.parse(code, filename=str(source.path))
     except SyntaxError as exc:
         logger.error(f"Could not parse code from {source}: {exc}")
         return
-    yield from conditional_imports(parsed_code)
-    yield from alternative_imports(parsed_code)
-    yield from dynamic_imports(parsed_code)
-    yield from docstring(parsed_code)
+    for node in ast.walk(parsed_code):
+        yield from conditional_imports(node)
+        yield from alternative_imports(node)
+        yield from dynamic_imports(node)
+        yield from docstring(node)
+        yield from regular_imports(node)
```

Comment:
Nice work extracting the `for node in ast.walk(parsed_code):` out of each helper function. 👍🏻

Reply:
+1 :+1: